TST suppress convergence warnings in coordinate descent

norbusan opened the pull request:
#### Reference Issues/PRs

closes #4780

#### What does this implement/fix? Explain your changes.

These warnings have been fixed by either increasing the number of
iterations of fitting, or by suppressing the warnings with pytest.warns

For parametrized tests a check is done so that only the failing tests
get the pytest.warns context manager, all other get a dummy context
manager that doesn't do anything.

#### Any other comments?

There is one instance where the warning does not disappear in
   test_linear_model_sample_weights_normalize_in_pipeline
(second context manager call), but I don't know why ...

norbusan committed:
Fix coordinate descent convergence warnings (Fixes #4780)

These warnings have been fixed by either increasing the number of
iterations of fitting, or by suppressing the warnings with pytest.warns

For parametrized tests a check is done so that only the failing tests
get the pytest.warns context manager, all other get a dummy context
manager that doesn't do anything.

There is one instance where the warning does not disappear in
   test_linear_model_sample_weights_normalize_in_pipeline
(second context manager call), but I don't know why ...

glemaitre commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -386,11 +390,19 @@ def test_model_pipeline_same_as_normalize_true(LinearModel, params):
-    y_pred_normalize = model_normalize.predict(X_test)
+    # in these two cases we get ConvergenceWarnings, but the tests
+    # do still succeed, so supress the warnings
+    if LinearModel in (ElasticNet, MultiTaskElasticNet) and params["l1_ratio"] == 0:
```
```suggestion
    if (
        LinearModel in (ElasticNet, MultiTaskElasticNet) and
        params["l1_ratio"] == 0
    ):
```

norbusan replied:
Thanks for the suggestions, changed to this layout in the last commit.

norbusan committed:
Fix long lines

glemaitre commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -535,8 +561,13 @@ def test_model_pipeline_same_dense_and_sparse(LinearModel, params):
 
-    model_dense.fit(X, y)
-    model_sparse.fit(X_sparse, y)
+    if (LinearModel is ElasticNet and params["l1_ratio"] == 0) or LinearModel is LassoCV:
```
```suggestion
    if (
        (LinearModel is ElasticNet and params["l1_ratio"] == 0) or
        LinearModel is LassoCV
    ):
```

norbusan replied:
Same as above, fixed with last commit. Thanks.

norbusan committed:
Better layout for long lines breaking

norbusan commented:
I just saw that other tests use the following style to suppress warnings:
```
with ignore_warnings(category=ConvergenceWarning):
```
Is there a preference for the one or the other?

glemaitre commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -724,14 +761,16 @@ def test_uniform_targets():
             y1.fill(y_values)
-            assert_array_equal(model.fit(X_train, y1).predict(X_test), y1)
+            with pytest.warns(ConvergenceWarning):
+                assert_array_equal(model.fit(X_train, y1).predict(X_test), y1)
```
It would be best to store the prediction and then make the check. I find the pattern of having an assert in `pytest.warns` a bit weird.

norbusan replied:
Something like
```
            with pytest.warns(ConvergenceWarning):
                pred = model.fit(X_train, y1).predict(X_test)
            assert_array_equal(pred, y1)
            assert_array_equal(model.alphas_, [np.finfo(float).resolution]*3)
```

norbusan replied:
Done that now.

glemaitre commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -458,7 +473,14 @@ def test_linear_model_sample_weights_normalize_in_pipeline(
-    reg_with_normalize.fit(X_train, y_train, sample_weight=sample_weight)
+
+    if model_name in ['Lasso', 'ElasticNet']:
+        ctxmgr = pytest.warns(ConvergenceWarning)
```
Apparently one of them is not raising the `ConvergenceWarning`.

norbusan replied:
Really? I just tried both alone, and in both cases new warnings appear.

norbusan replied:
Ahh indeed, in some of the CI runs it does indeed not error out ...

I guess that answers my other quesion that one should use `with ignore_warnings(category=ConvergenceWarning):`

glemaitre commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -458,7 +473,14 @@ def test_linear_model_sample_weights_normalize_in_pipeline(
+    if model_name in ['Lasso', 'ElasticNet']:
+        ctxmgr = pytest.warns(ConvergenceWarning)
+    else:
+        ctxmgr = contextlib.suppress()
```
It might be good to have a comment to mention that `supress()` without parameters will not suppress warnings :)

norbusan replied:
Indeed! I would liked to use Python >= 3.7 contextlib.nullcontext, but that is not supported on 3.6 it seems. Will update in a minute.

glemaitre replied:
Oh indeed, we will support only `Python 3.7` in version 1.0 so we could be proactive. I don't remember if we move all our CI from 3.6 to 3.7

norbusan replied:
For now I added a comment to the first instance of `suppress()`, and mentioned also Python 3.7 there. If all is switched we can use nullcontext (which would somehow be *nicer*)

cmarmo replied:
@glemaitre all the CI has been moved to python 3.7.

norbusan replied:
Great, so I have switched to use `nullcontext()` - let us see how it goes ;-)

norbusan committed:
Factor out prediction computation, add comment about suppress()

norbusan committed:
Merge remote-tracking branch 'upstream/main' into issue4780

norbusan committed:
various updates concerning warnings

- use ignore_warnings instead of pytest.warns since depending on
  the used Python system the warning might not be issued
- deal with multiple warnings in a single statement
- use nullcontext() instead of suppress() since we have moved to
  Python 3.7 as dependency and CI testing system

ogrisel commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -111,7 +113,10 @@ def test_lasso_zero():
     y = [0, 0, 0]
-    clf = Lasso(alpha=0.1).fit(X, y)
+    # _cd_fast.pyx tests for gap < tol, but here we get 0.0 < 0.0
+    # should probably be changed to gap <= tol ?
```
Indeed, please feel free to open a dedicated follow-up PR once this one is merged.

ogrisel replied:
Actually I think this should be fixed as part of this PR to simplify it.

ogrisel replied:
Also once this is done, could you please update this PR to add inline comments to explain why we expect a given warning each time we ignore it in the tests?

norbusan replied:
Thanks for the comments, I will push a change for `_cd_fast.pyx` but want to do some testing with that (first non-python only change).

Concerning the comments: that is a good point. My idea is that what we are checking here is that two different methods to do a fitting (pipeline or intern scaling for example) result in the same fitting, not whether the fitting actually converges.

But I will  **try** to check for each case to find an explanation ...

norbusan replied:
@ogrisel Ok, that turned out to be a bit more tricky and takes time (at least for me): If I change the code in `_cd_fast.pyx` to do
```
@@ -819,7 +819,7 @@ def enet_coordinate_descent_multi_task(
                 gap += l1_reg * l21_norm - const * ry_sum + \
                      0.5 * l2_reg * (1 + const ** 2) * (w_norm ** 2)
 
-                if gap < tol:
+                if gap <= tol:
                     # return if we reached desired tolerance
                     break
         else:
```
then suddenly a warning that *should* (?) be there disappeared: the test `test_enet_coordinate_descent` in `test_coordinate_descent.py` has the following comment:
```
    """Test that a warning is issued if model does not converge"""
```
and is tested against `Lasso` (`precompute = True | False`) and against `MultiTaskLasso` (interestingly, two times the same test), and now suddenly `MultiTaskLasso` converges while `Lasso` does not.

The code that is run is basically
```
X = np.ones((5, 2)) * 1e50
y = np.ones((5,2))
MultiTaskLasso(max_iter=2).fit(X,y)
```
Since this boils down to looking at what is going on in `_cd_fast.pyx`, function `enet_coordinate_descent_multi_task` I need a bit of time to understand how that works.

If any of you have an insight at hand, please let me know.

norbusan replied:
BTW, I dropped the duplication of `MultiTaskLasso` in the last commit pushed, but that is cosmetics.

ogrisel replied:
Can you please push the change, either as part of this PR or in a new draft PR to see what is failing with specific error message and so that we can have a more focused discussion on particular failures caused by this change?

norbusan replied:
Hi @ogrisel I pushed the `gap <= tol` change to the PR (and a typo fix).

glemaitre commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -111,7 +113,10 @@ def test_lasso_zero():
-    clf = Lasso(alpha=0.1).fit(X, y)
+    # _cd_fast.pyx tests for gap < tol, but here we get 0.0 < 0.0
+    # should probably be changed to gap <= tol ?
+    with ignore_warnings(category=ConvergenceWarning):
```
I am wondering if we could decorate the function with a `pytest.mark.filterwarning` instead?

norbusan replied:
Yes, in principle we could decorate the functions that way to simplify - probably a style questions - decorations versus explicit code.

But first I want to see what is going on with `gap <= tol` in `_cd_fast.pyx`, see above.

glemaitre commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -386,10 +391,39 @@ def test_model_pipeline_same_as_normalize_true(LinearModel, params):
-    model_normalize.fit(X_train, y_train)
-    y_pred_normalize = model_normalize.predict(X_test)
+    # The following fit calls might produce two different warnings:
+    # - UserWarning:Coordinate descent with l1_reg=0 may lead
+    #               to unexpected results and is discouraged.
```
```suggestion
    # - UserWarning: Coordinate descent with l1_reg=0 may lead
    #   to unexpected results and is discouraged.
```

norbusan replied:
Fixed with latest push

glemaitre commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -1059,7 +1119,8 @@ def test_check_input_false():
     # pass anyway
     X = check_array(X, order='F', dtype='float32')
-    clf.fit(X, y, check_input=False)
+    with ignore_warnings(category=ConvergenceWarning):
```
we could decorate the function with a decorator `pytest.mark.filterwarnings`

glemaitre commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -724,14 +780,18 @@ def test_uniform_targets():
         for y_values in (0, 5):
             y1.fill(y_values)
-            assert_array_equal(model.fit(X_train, y1).predict(X_test), y1)
+            with ignore_warnings(category=ConvergenceWarning):
```
since we ignore the warning all the time, we could decorate the test function as well.

glemaitre reviewed (commented):
I think that sometimes, we could use `pytest.mark.filterwarning` instead of `ignore_warning`.

norbusan committed:
Style fix and duplication fix

ogrisel commented on sklearn/linear_model/tests/test_coordinate_descent.py:
```diff
@@ -111,7 +113,10 @@ def test_lasso_zero():
     X = [[0], [0], [0]]
     y = [0, 0, 0]
-    clf = Lasso(alpha=0.1).fit(X, y)
+    # _cd_fast.pyx tests for gap < tol, but here we get 0.0 < 0.0
```
Actually I do not understand why it would be the case in this test. By default the tolerance is 1e-4:

```python
>>> from sklearn.linear_model import Lasso
>>> Lasso().tol
0.0001
```

This test does not override the default. The same is true for `MultiTaskLasso`.

ogrisel replied:
Ah ok I understand, this is because we rescale the tolerance by a data dependent quantity which is 0 in this case.

norbusan replied:
That is indeed sub-optimal - `y == 0` are valid options. What about the following change to keep `tol` as is if it would become `0` (similar code for all the different coordinate descent function instances):
```
--- a/sklearn/linear_model/_cd_fast.pyx
+++ b/sklearn/linear_model/_cd_fast.pyx
@@ -672,6 +672,7 @@ def enet_coordinate_descent_multi_task(
     cdef floating W_ii_abs_max
     cdef floating gap = tol + 1.0
     cdef floating d_w_tol = tol
+    cdef floating temp_tol = tol
     cdef floating R_norm
     cdef floating w_norm
     cdef floating ry_sum
@@ -704,7 +705,9 @@ def enet_coordinate_descent_multi_task(
                           &R[0, jj], 1)
 
         # tol = tol * linalg.norm(Y, ord='fro') ** 2
-        tol = tol * _nrm2(n_samples * n_tasks, Y_ptr, 1) ** 2
+        temp_tol = tol * _nrm2(n_samples * n_tasks, Y_ptr, 1) ** 2
+        if temp_tol > 0:
+            tol = temp_tol
 
         for n_iter in range(max_iter):
             w_max = 0.0
@@ -819,7 +822,7 @@ def enet_coordinate_descent_multi_task(
                 gap += l1_reg * l21_norm - const * ry_sum + \
                      0.5 * l2_reg * (1 + const ** 2) * (w_norm ** 2)
 
-                if gap <= tol:
+                if gap < tol:
                     # return if we reached desired tolerance
                     break
         else:
```
Then we could also return to the previous (`gap < tol`).

norbusan committed:
Use gap <= tol for check

cmarmo commented:
Hi @norbusan, thank you for your work so far! Are you still interested in finalizing this pull request? The [list of the tests with convergence issues](https://github.com/scikit-learn/scikit-learn/issues/4780#issuecomment-671326256) slightly changed.

adrinjalali committed:
Merge remote-tracking branch 'upstream/main' into issue4780

adrinjalali committed:
remove failing convergence warning test

adrinjalali commented:
I gave this a fresh start @ogrisel @glemaitre . Had to remove the `MultiTaskLasso` test case for convergence warning since I couldn't create a warning with this PR.

glemaitre reviewed (approved):
LGTM
